receiver/splunkhecreceiver/splunk_to_logdata.go

```diff
@@ -0,0 +1,162 @@
+			logRecord.Body().SetIntVal(value)
+		} else if value, ok := event.Event.(float64); ok {
+			logRecord.Body().SetDoubleVal(value)
+		} else if value, ok := event.Event.(bool); ok {
+			logRecord.Body().SetBoolVal(value)
+		} else if value, ok := event.Event.(map[string]interface{}); ok {
+			mapValue, err := convertToAttributeMap(logger, value)
+			if err != nil {
+				return lrs, err
+			}
+			logRecord.Body().SetMapVal(mapValue)
+		} else if value, ok := event.Event.([]interface{}); ok {
+			arrValue, err := convertToArrayVal(logger, value)
+			if err != nil {
+				return lrs, err
+			}
+			logRecord.Body().SetArrayVal(arrValue)
+		}
+
+		// Splunk timestamps are in seconds so convert to nanos by multiplying
+		// by 1 billion.
+		logRecord.SetTimestamp(pdata.TimestampUnixNano(event.Time * 1e9))
+
+		lr.Resource().InitEmpty()
+		attrs := lr.Resource().Attributes()
+		attrs.InitEmptyWithCapacity(3)
+		attrs.InsertString(conventions.AttributeHostHostname, event.Host)
+		attrs.InsertString(conventions.AttributeServiceName, event.Source)
+		attrs.InsertString(splunk.SourcetypeLabel, event.SourceType)
+		//TODO consider setting the index field as well for pass through scenarios.
+		keys := make([]string, 0, len(event.Fields))
+		for k := range event.Fields {
+			keys = append(keys, k)
+		}
+		sort.Strings(keys)
+		for _, key := range keys {
+			val := event.Fields[key]
+			attrValue, err := convertInterfaceToAttributeValue(logger, val)
+			if err != nil {
+				return lrs, err
+			}
+			logRecord.Attributes().Insert(key, attrValue)
+		}
+		ill := pdata.NewInstrumentationLibraryLogs()
+		ill.InitEmpty()
+		ill.Logs().Append(logRecord)
+		lr.InstrumentationLibraryLogs().Append(ill)
+	}
+
+	return lrs, nil
+}
+
+func convertInterfaceToAttributeValue(logger *zap.Logger, originalValue interface{}) (pdata.AttributeValue, error) {
+	if originalValue == nil {
+		return pdata.NewAttributeValueNull(), nil
+	} else if value, ok := originalValue.(string); ok {
+		return pdata.NewAttributeValueString(value), nil
+	} else if value, ok := originalValue.(int64); ok {
+		return pdata.NewAttributeValueInt(value), nil
+	} else if value, ok := originalValue.(float64); ok {
+		return pdata.NewAttributeValueDouble(value), nil
+	} else if value, ok := originalValue.(bool); ok {
+		return pdata.NewAttributeValueBool(value), nil
+	} else if value, ok := originalValue.(map[string]interface{}); ok {
+		mapContents, err := convertToAttributeMap(logger, value)
+		if err != nil {
+			return pdata.NewAttributeValueNull(), err
+		}
+		mapValue := pdata.NewAttributeValueMap()
+		mapValue.SetMapVal(mapContents)
+		return mapValue, nil
+	} else if value, ok := originalValue.([]interface{}); ok {
+		arrValue := pdata.NewAttributeValueArray()
+		arrContents, err := convertToArrayVal(logger, value)
+		if err != nil {
+			return pdata.NewAttributeValueNull(), err
+		}
+		arrValue.SetArrayVal(arrContents)
+		return arrValue, nil
+	} else {
```

Comment:
This looks similar to the above code that deals with `Body`. Is it possible to have only one copy of this code? (I am not sure, maybe not, just checking).

Reply:
No, to my chagrin. I can set the `Body` value but I cannot set it itself.

Reply:
That might be a good RFE for core otel - add a `.setVal(attrValue AttributeValue)` method to Body.

Reply:
Actually, looks like I can use CopyTo to get the job done. Fixed.